patroni/postgresql.py

```diff
@@ -223,24 +223,72 @@ def sync_from_leader(self, leader):
-        return self.create_replica(r, env) == 0
+        ret = self.create_replica(leader, env) == 0
+        ret and self.delete_trigger_file()
+        return ret
 
     @staticmethod
     def build_connstring(conn):
-        return "host={host} port={port} user={user}".format(**conn)
+        """
+        >>> Postgresql.build_connstring({'host': '127.0.0.1', 'port': '5432'}) == 'host=127.0.0.1 port=5432 '
+        True
+        """
+        mconn = ""
+        for param, val in sorted(conn.items()):
+            mconn = mconn + "{0}={1} ".format(param, val)
+
+        return mconn
+
+    def create_replica(self, leader, env):
+        # create the replica according to the replica_method
+        # defined by the user.  this is a list, so we need to
+        # loop through all methods the user supplies
+        connstring = leader.conn_url
+        # get list of replica methods from config
+        replica_list = self.config.get('create_replica_method', 'basebackup')
+        replica_methods = [rm.strip() for rm in replica_list.split(',')]
+        # go through them in priority order
+        ret = 1
+        for replica_method in replica_methods:
+            # if the method is basebackup, then use the built-in
+            if replica_method == "basebackup":
+                ret = self.basebackup(leader, env)
+                if ret == 0:
+                    # if basebackup succeeds, exit with success
+                    break
+            else:
+                cmd = replica_method
+                # user-defined method; check for configuration
+                # not required, actually
+                if replica_method in self.config:
+                    method_config = self.config[replica_method].copy()
+                    # look to see if the user has supplied a full command path
+                    # if not, use the method name as the command
+                    if "command" in method_config:
+                        cmd = method_config["command"]
+                        # remove the command and turn it into a shlex set
+                        del method_config["command"]
+                    # add the default parameters
+                    method_config.update({"scope": self.scope,
+                                          "role": "replica",
+                                          "datadir": self.data_dir,
+                                          "connstring": connstring})
+                    params = ["--{0}={1}".format(arg, val) for arg, val in method_config.items()]
+                else:
+                    method_config = {"scope": self.scope,
+                                     "role": "replica",
+                                     "datadir": self.data_dir,
+                                     "connstring": connstring}
+                try:
+                    # call script with the full set of parameters
+                    ret = subprocess.call(shlex.split(cmd) + params, env=env)
```

Comment:
`params` could be not defined here, because it is not declared in `else` block from line 277